File: 프로그래머스/연습문제/Lv2/혼자 놀기의 달인/혼자 놀기의 달인.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

int solution(vector<int> cards) {
    vector<int> scores; // 점수들
    vector<bool> visit(cards.size(), false);
    
    for(int i=0;i <cards.size(); i++){
        if(!visit[i]){
            int score = 1;
            visit[i] = true;
            int idx = cards[i]-1;
            
            while(1){
                if(visit[idx]) break;
                
                score++;
                visit[idx] = true;
                idx = cards[idx]-1;
            }
            
            scores.push_back(score);
        }
    }
    
    sort(scores.begin(), scores.end(), greater<int>());
    
    if(scores.size() <= 1) return 0;
    else return scores[0]*scores[1];
}

File: 프로그래머스/연습문제/Lv2/혼자 놀기의 달인/혼자 놀기의 달인.py
def solution(cards):
    scores = []
    visit = [False for i in range(len(cards))]

    for i in range(len(cards)):
        if not visit[i]:
            score = 1
            visit[i] = True
            idx = cards[i]-1

            while True:
                if visit[idx]: break

                score += 1
                visit[idx] = True
                idx = cards[idx]-1

            scores.append(score)

    scores = sorted(scores, reverse=True)

    if len(scores) <= 1: return 0
    else: return scores[0]*scores[1]
